File: copilot-bot/src/split-send.ts
import { MessageFlags, type PartialTextBasedChannelFields, TextDisplayBuilder } from "discord.js";

const MAX_LENGTH = 2000;

// Markdown patterns that come in pairs and shouldn't be split
const MARKDOWN_PAIRS = [
    { pattern: /```/g, name: "code block" }, // Code blocks
    { pattern: /``/g, name: "inline code" }, // Double backtick inline code
    { pattern: /`/g, name: "inline code" }, // Single backtick inline code
    { pattern: /\*\*/g, name: "bold" }, // Bold
    { pattern: /__/g, name: "underline bold" }, // Underline/bold
    { pattern: /\|\|/g, name: "spoiler" }, // Spoilers
    { pattern: /~~/g, name: "strikethrough" }, // Strikethrough
    { pattern: /(?<!\*)\*(?!\*)/g, name: "italic" }, // Italic (single asterisk, not part of bold)
    { pattern: /(?<!_)_(?!_)/g, name: "italic" }, // Italic (single underscore, not part of underline)
];

/**
 * Counts occurrences of a pattern in a string
 */
function countMatches(text: string, pattern: RegExp): number {
    const matches = text.match(pattern);
    return matches ? matches.length : 0;
}

/**
 * Checks if a chunk has unclosed markdown formatting
 */
function hasUnclosedMarkdown(chunk: string): { unclosed: boolean; pattern?: RegExp } {
    for (const { pattern } of MARKDOWN_PAIRS) {
        const count = countMatches(chunk, pattern);
        // If odd number of occurrences, the formatting is unclosed
        if (count % 2 !== 0) {
            return { unclosed: true, pattern };
        }
    }
    return { unclosed: false };
}

/**
 * Finds the last safe split point that doesn't break markdown formatting
 */
function findSafeSplitPoint(text: string, maxLength: number): number {
    // First, try to split at a newline for cleaner breaks
    const lastNewline = text.lastIndexOf("\n", maxLength);
    if (lastNewline > maxLength * 0.5) {
        // Check if splitting here would break markdown
        const chunk = text.slice(0, lastNewline);
        if (!hasUnclosedMarkdown(chunk).unclosed) {
            return lastNewline + 1; // Include the newline in the first chunk
        }
    }

    // Try to find a safe split point by checking progressively shorter chunks
    // Prioritize splitting at paragraph breaks (double newlines)
    const lastParagraph = text.lastIndexOf("\n\n", maxLength);
    if (lastParagraph > maxLength * 0.3) {
        const chunk = text.slice(0, lastParagraph);
        if (!hasUnclosedMarkdown(chunk).unclosed) {
            return lastParagraph + 2;
        }
    }

    // Try splitting at code block boundaries (```)
    const codeBlockRegex = /```[\s\S]*?```/g;
    let lastCodeBlockEnd = 0;
    let match: RegExpExecArray | null;
    const textToSearch = text.slice(0, maxLength);

    // biome-ignore lint/suspicious/noAssignInExpressions: it works
    while ((match = codeBlockRegex.exec(textToSearch)) !== null) {
        const endPos = match.index + match[0].length;
        if (endPos <= maxLength) {
            lastCodeBlockEnd = endPos;
        }
    }

    if (lastCodeBlockEnd > maxLength * 0.3) {
        return lastCodeBlockEnd;
    }

    // Fall back to finding a safe point by checking each potential split
    // Work backwards from maxLength, preferring spaces/newlines
    for (let i = maxLength; i > maxLength * 0.3; i--) {
        const char = text[i];
        if (char === "\n" || char === " ") {
            const chunk = text.slice(0, i);
            if (!hasUnclosedMarkdown(chunk).unclosed) {
                return char === "\n" ? i + 1 : i;
            }
        }
    }

    // If we couldn't find a safe markdown break point, just split at a space
    const lastSpace = text.lastIndexOf(" ", maxLength);
    if (lastSpace > maxLength * 0.3) {
        return lastSpace;
    }

    // Last resort: hard split at maxLength
    return maxLength;
}

/**
 * Splits a message into chunks respecting markdown formatting boundaries
 */
function splitMessageSafely(message: string): string[] {
    const chunks: string[] = [];
    let remaining = message;

    while (remaining.length > 0) {
        if (remaining.length <= MAX_LENGTH) {
            chunks.push(remaining);
            break;
        }

        const splitPoint = findSafeSplitPoint(remaining, MAX_LENGTH);
        const chunk = remaining.slice(0, splitPoint).trim();

        if (chunk.length > 0) {
            chunks.push(chunk);
        }

        remaining = remaining.slice(splitPoint).trim();
    }

    return chunks;
}

async function send(message: string, channel: PartialTextBasedChannelFields) {
    const text = new TextDisplayBuilder().setContent(message);
    await channel.send({ components: [text], flags: MessageFlags.IsComponentsV2 });
}

export async function splitAndSendAsComponents(message: string, channel: PartialTextBasedChannelFields) {
    if (message.length > MAX_LENGTH) {
        const chunks = splitMessageSafely(message);
        for (const chunk of chunks) {
            await send(chunk, channel);
        }
    } else {
        await send(message, channel);
    }
}

File: copilot-bot/splitsend.go
package main

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/bwmarrin/discordgo"
)

const maxMessageLength = 2000

// markdownPatterns lists patterns that come in pairs.
// An odd count means unclosed formatting that we shouldn't split inside.
var markdownPatterns = []*regexp.Regexp{
	regexp.MustCompile("```"),   // Code blocks
	regexp.MustCompile("``"),    // Double backtick
	regexp.MustCompile("`"),     // Inline code
	regexp.MustCompile(`\*\*`),  // Bold
	regexp.MustCompile("__"),    // Underline/bold
	regexp.MustCompile(`\|\|`),  // Spoilers
	regexp.MustCompile("~~"),    // Strikethrough
}

func hasUnclosedMarkdown(chunk string) bool {
	for _, pattern := range markdownPatterns {
		matches := pattern.FindAllString(chunk, -1)
		if len(matches)%2 != 0 {
			return true
		}
	}
	return false
}

func findSafeSplitPoint(text string, maxLen int) int {
	end := min(len(text), maxLen)

	// Try paragraph break (double newline)
	if idx := strings.LastIndex(text[:end], "\n\n"); idx > maxLen*3/10 {
		if !hasUnclosedMarkdown(text[:idx]) {
			return idx + 2
		}
	}

	// Try single newline
	if idx := strings.LastIndex(text[:end], "\n"); idx > maxLen/2 {
		if !hasUnclosedMarkdown(text[:idx]) {
			return idx + 1
		}
	}

	// Walk backwards looking for a safe whitespace break
	for i := end; i > maxLen*3/10; i-- {
		ch := text[i-1]
		if ch == ' ' || ch == '\n' {
			if !hasUnclosedMarkdown(text[:i]) {
				return i
			}
		}
	}

	// Fallback: split at last space
	if idx := strings.LastIndex(text[:end], " "); idx > maxLen*3/10 {
		return idx
	}

	// Last resort: hard cut
	return maxLen
}

func splitMessage(message string) []string {
	var chunks []string
	remaining := message

	for len(remaining) > 0 {
		if len(remaining) <= maxMessageLength {
			chunks = append(chunks, remaining)
			break
		}

		splitPoint := findSafeSplitPoint(remaining, maxMessageLength)
		chunk := strings.TrimSpace(remaining[:splitPoint])
		if chunk != "" {
			chunks = append(chunks, chunk)
		}
		remaining = strings.TrimSpace(remaining[splitPoint:])
	}

	return chunks
}

// SplitAndSend splits a long message into Discord-safe chunks and sends them.
// It returns the first error encountered — callers decide how to handle it.
func SplitAndSend(s *discordgo.Session, channelID, message string) error {
	if len(message) <= maxMessageLength {
		_, err := s.ChannelMessageSend(channelID, message)
		return err
	}

	for _, chunk := range splitMessage(message) {
		if _, err := s.ChannelMessageSend(channelID, chunk); err != nil {
			return fmt.Errorf("sending message chunk: %w", err)
		}
	}
	return nil
}
